### ISSUE TITLE
Publish to npm?

### ISSUE THREAD

**Issue by ericelliott** [github/fetch#9]
[Because](https://vimeo.com/89258863) npm is the closest thing we have to a standard module repository for JavaScript (including the client-side). It's the largest module registry of its kind for any language. There are about 1/5th as many Bower modules as npm modules.

If that's not convincing enough: http://browserify.org/articles.html

**Comment by qubyte** [comment]
@rauchg Expanding on that, I wrote an [ugly little wrapper](https://github.com/qubyte/fetch-ponyfill) around this one that stops it touching `window`. It is a function that takes `Promise` and `XMLHttpRequest` constructors as arguments to return a `fetch` function. You could feed it `node-xmlhttprequest` to get the result you want in Node. That said, it would be better for this library just not to touch `window`. The wrapper uses a `new Function`, which I'm not happy about.

**Comment by rauchg** [comment]
I dig it!

**Comment by rauchg** [comment]
@qubyte 

To expand, I think it's a very good idea to expose a function that returns the `fetch` polyfill, especially in Node.JS.

The main reason is that we'd have to start adding non-standard properties to the object passed to `fetch` for the inevitable need of more flexibility and configurability. Things like custom SSL options, a different `Agent` object, etc.

From a practicality standpoint, however, I wonder if it makes sense to pass the `XMLHttpRequest` instance though.

One reason is that `node-xmlhttprequest` already exhibits the lack of customizability mentioned above. I reported it to them [here](https://github.com/driverdan/node-XMLHttpRequest/issues/44), because we use it extensively in engine.io-client and socket.io-client.

Their interim solution is to configure a single `XMLHttpRequest` class globally, which is not ideal if several instances with different options need to co-exist.

Maybe it's better to abandon the idea of sharing the code that depends on `XMLHttpRequest`, and internally use `mikeal/request` for node. The options passed to `require('fetch')()` are then passed onto `mikeal/request`.

All in all, what's important is that `require('fetch')()` has the exact same API as the spec, in both environments.

**Comment by ericelliott** [comment]
> Maybe it's better to abandon the idea of sharing the code that depends on XMLHttpRequest, and internally use mikeal/request for node.

:+1: 

> All in all, what's important is that require('fetch')() has the exact same API as the spec, in both environments.

Agreed.

**Comment by josh** [comment]
> The main reason is that we'd have to start adding non-standard properties to the object passed to fetch for the inevitable need of more flexibility and configurability. Things like custom SSL options, a different Agent object, etc.

Exactly why all this would be better done in a separate library that targets node.js directly rather than building on top of XHR.

**Comment by josh** [comment]
I think I'm just going to close this issue out.

**This** repo is going to remain a browser polyfill assuming a browser environment.

Definitely checkout https://github.com/matthew-andrews/isomorphic-fetch and maybe they'll be a node-fetch build on top of node's http lib. I'd be happy to link those up in the readme.

Thanks.

**Comment by domenic** [comment]
I still think this exact repo should be published to npm. npm is a great way to install polyfills.

**Comment by ericelliott** [comment]
@domenic :+1:

**Comment by josh** [comment]
We'll definitely need an alt name on npm since fetch is taken. Hopefully something makes it clear that `require('fetch')` will **not** will in node.js.

**Comment by josh** [comment]
Also pro if someone could setup travis' npm releasing http://docs.travis-ci.com/user/deployment/npm/ so I don't have to be bothered forgetting the extra step.

**Comment by matthew-andrews** [comment]
@josh, I've made a tool [npm-prepublish](https://github.com/matthew-andrews/npm-prepublish) that allows you to use GitHub releases & npm publish via Travis together seamlessly.

the way it works is you first **remove** the `version` property from `package.json` then run `npm install --save-dev npm-prepublish` and set up the `travis.yml` file so that it will do an `npm publish` with `npm-prepublish` as a `before_deploy` script (which puts the version number from the tag into the `package.json`) [all this is documented poorly — sorry, I will improve — on `npm-prepublish`'s repo].

Basically, you don't have to do the irritating/easy to forget `npm version blah && git push && git push --tags && npm publish` dance.ust go to github, create a github release and the rest will just work.

Here's a worked example:-
https://github.com/Financial-Times/next-express/compare/v2.2.1...v2.2.2

**Comment by hawkrives** [comment]
@matthew-andrews Travis also has a feature to push a new NPM release whenever there's a new git tag automatically. 

I should be able to submit a pull request for that later today.

**Comment by matthew-andrews** [comment]
@hawkrives I use that.  The problem is if you use the default Travis/npm approach you can't use GitHub releases otherwise the version number in `package.json` and the version in the `tag` drift out of sync, you **must** use `npm version` on the command line to update the `version` property in `package.json`.

My module eliminates this unnecessary step and requirement to release via the command line.

**Comment by hawkrives** [comment]
Ah.

**Comment by necolas** [comment]
the `fetch` package name was offered to you -> https://github.com/github/fetch/issues/9#issuecomment-59704339

**Comment by matthew-andrews** [comment]
Thanks @necolas for the reminder… I've put a note in the PR: https://github.com/github/fetch/pull/57

**Comment by matthew-andrews** [comment]
@hawkrives I owe you an apology.  It turns out they were happy with the edit-package.json-and-bower.json-update-the-version-number-git-tag-git-push---tags workflow after all and that the changes you described were exactly what they wanted.  I'm sorry for discouraging you from making a pull request that did that.

**Comment by hawkrives** [comment]
@matthew-andrews No worries. Yours cleared up some good ambiguities; I apologize for being slightly short with you.

Also, that's an epic hyphenated word in that post. Keep up the good work!

**Comment by tracker1** [comment]
@josh how about `browser-fetch-polyfill` or `browser-polyfill-fetch`?

Being explicit in that it's a browser polyfill for fetch in the naming... regarding publishing to npm.

**Comment by matthew-andrews** [comment]
@tracker1 it's published now as `whatwg-fetch`. We could publish it twice under two different names but of that works for you it would be simpler to keep it as it is :).

**Comment by ericelliott** [comment]
Better to have a single canonical name. `whatwg-fetch` is fine, IMO.

**Comment by bitinn** [comment]
For people interested in a node.js implementation:

https://github.com/bitinn/node-fetch

**Comment by rauchg** [comment]
@bitinn good stuff

**Comment by josh** [comment]
@bitinn nice! I was waiting for someone to write just that. Mind opening a PR to cross link it from our README? I totally think anyone wanting to use fetch in just node.js should be using your lib instead.
